File: code-examples/java/queue/src/queueDSA.java
import java.util.Arrays;

public class queueDSA {
    private int maxSize;
    private int[] queueArray;
    private int front;
    private int rear;
    private int dataCount;

    //constructor
    public queueDSA (int size){
        maxSize = size;
        queueArray = new int[maxSize];
        front = 0;
        rear = 0;
        dataCount=0;
    }

    //method to check if the queue is empty
    public boolean isEmpty(){
        return dataCount == 0;
    }

    //method to check if the queue is full
    public boolean isFull(){
        return dataCount == maxSize;
    }

    //method to add to the queue (adds element at the rear)
    public void enqueue(int number){
        if(isFull()){
            System.out.println("Queue is Full!");
        }else{
            queueArray[rear] = number;
            dataCount++;

            rear = (rear+1) % maxSize; //use to wrap the front and rear back to 0 when they hit maxSize
        }
    }

    //method to remove element from the queue (removes from the front)
    public void dequeue(){
        if(isEmpty()){
            System.out.println("Queue is Empty!");
        }else{

            dataCount--;
            System.out.println("Dequeue Value: "+queueArray[front]);
            front = (front+1) % maxSize;
        }
    }

    //method to peek the front element
    public void peek(){
        System.out.println(queueArray[front]);
    }

    //method to get the currant size of the queue
    public void size(){
        System.out.println(dataCount);
    }

    /*
    Debug codes
    //method to print Array
    public void printArray(){
        System.out.println("The Array: "+(Arrays.toString(queueArray)));
    }

    public void peek(){
        System.out.println("Front: "+front);
        System.out.println("Rear: "+rear);
        System.out.println("Data Count: "+dataCount);

    }
     */
}

File: code-examples/c++/queue/src/QueueDSA.cpp
#include "QueueDSA.h"
#include <iostream>

QueueDSA::QueueDSA(int size){
    maxSize = size;
    queueArray = new int[maxSize];
    front = 0;
    rear = 0;
    dataCount = 0;
}

QueueDSA::~QueueDSA(){
    delete[] queueArray;
}

bool QueueDSA::isEmpty(){
    return dataCount == 0;
}

bool QueueDSA::isFull(){
    return dataCount == maxSize;
}

void QueueDSA::enqueue(int number){
    if(isFull()){
        std::cout<<"Queue is Full!!!"<<std::endl;
    }else{
        queueArray[rear] = number;
        rear = (rear+1) % maxSize;
        dataCount++;
    }
}

void QueueDSA::dequeue(){
    if(isEmpty()){
        std::cout<<"Queue is Empty!!!"<<std::endl;
    }else{
        std::cout<<"Dequeue: "<<queueArray[front]<<std::endl;
        front = (front+1) % maxSize;
        dataCount--;
    }
}

void QueueDSA::sizeOfQueue(){
    std::cout<<"Size of the Queue: "<<dataCount<<std::endl;
}

void QueueDSA::peek(){
    if(isEmpty()){
        std::cout<<"Queue is Empty!!!"<<std::endl;
    }else{
        std::cout<<"Peek: "<<queueArray[front]<<std::endl;
    }
}

/* Debuggin code
void QueueDSA::printQueue(){
    if(isEmpty()){
        std::cout<<"Queue is Empty!!!"<<std::endl;
    }else{

        int index = front;
        
        for(int i=0; i<dataCount; i++){
            std::cout<<queueArray[index]<<std::endl;
            index = (index+1) % maxSize;
        }
        
        std::cout<<"The second array starts!"<<std::endl;

        for (int i = 0; i < maxSize; i++)
        {
            std::cout<<queueArray[i]<<std::endl;
        }
        
    }
}*/